xbmc/platform/android/activity/JNIXBMCRecognitionListener.cpp

```diff
@@ -0,0 +1,94 @@
+/*
+ *  Copyright (C) 2022 Team Kodi
+ *  This file is part of Kodi - https://kodi.tv
+ *
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *  See LICENSES/README.md for more information.
+ */
+
+#include "JNIXBMCRecognitionListener.h"
+
+#include "CompileInfo.h"
+#include "ServiceBroker.h"
+#include "dialogs/GUIDialogKaiToast.h"
+#include "guilib/GUIComponent.h"
+#include "guilib/GUIWindow.h"
+#include "guilib/GUIWindowManager.h"
+#include "guilib/LocalizeStrings.h"
+#include "messaging/ApplicationMessenger.h"
+#include "utils/log.h"
+
+#include <androidjni/Context.h>
+#include <androidjni/jutils-details.hpp>
+
+using namespace jni;
+
+static std::string s_className =
+    std::string(CCompileInfo::GetClass()) + "/interfaces/XBMCRecognitionListener";
+
+CJNIXBMCRecognitionListener::CJNIXBMCRecognitionListener() : CJNIBase(s_className)
+{
+  m_object = new_object(CJNIContext::getClassLoader().loadClass(GetDotClassName(s_className)));
+  m_object.setGlobal();
+
+  add_instance(m_object, this);
+}
+
+void CJNIXBMCRecognitionListener::RegisterNatives(JNIEnv* env)
+{
+  jclass cClass = env->FindClass(s_className.c_str());
+  if (cClass)
+  {
+    JNINativeMethod methods[] = {
+        {"_onReadyForSpeech", "(Landroid/os/Bundle;)V",
+         reinterpret_cast<void*>(&CJNIXBMCRecognitionListener::_onReadyForSpeech)},
+        {"_onError", "(I)V", reinterpret_cast<void*>(&CJNIXBMCRecognitionListener::_onError)},
+        {"_onResults", "(Landroid/os/Bundle;)V",
+         reinterpret_cast<void*>(&CJNIXBMCRecognitionListener::_onResults)},
+    };
+
+    env->RegisterNatives(cClass, methods, sizeof(methods) / sizeof(methods[0]));
+  }
+}
+
+CJNIXBMCRecognitionListener::~CJNIXBMCRecognitionListener()
+{
+  remove_instance(this);
+}
+
+void CJNIXBMCRecognitionListener::_onReadyForSpeech(JNIEnv* env, jobject thiz, jobject bundle)
+{
+  CGUIDialogKaiToast::QueueNotification(CGUIDialogKaiToast::Info, g_localizeStrings.Get(39177),
+      g_localizeStrings.Get(39179));
+
+  CLog::Log(LOGDEBUG, "CJNIXBMCRecognitionListener onReadyForSpeech()");
+}
+
+void CJNIXBMCRecognitionListener::_onError(JNIEnv* env, jobject thiz, jint i)
+{
+  if (i == CJNISpeechRecognizer::ERROR_NO_MATCH)
+    CGUIDialogKaiToast::QueueNotification(CGUIDialogKaiToast::Info, g_localizeStrings.Get(39177),
+        g_localizeStrings.Get(39180));
+  else
+    CGUIDialogKaiToast::QueueNotification(CGUIDialogKaiToast::Info, g_localizeStrings.Get(39177),
+        g_localizeStrings.Get(39181));
```

Comment:
Dependency to high level GUI layer here is not okay. Either you use messaging to signal the error or you need to drop the feature to interact with the user on errors.

Reply:
Most probably you need to define and implement a new GUI_MSG_USER.

Reply:
These info toast are displayed to the user to give him some feedback on the speech recognition process. It's not possible to interact with these errors as the listening process has ended.

These messages only appear when the Speech Recognizer has not been able to match what it has heard or there has been an error (connection error...).

Reply:
I understand the purpose of the toasts, but from software architectural standpoint your implementation is wrong.  You are in a lower non gui layer, which must a) not directly call into upper gui layer of Kodi and b) you are in an arbitrary thread and Kodi guilib Must only be called from kodis main thread. We have AppMessenger to deal with both problems.

Reply:
Ok, I'll make the changes to send the messages to ```CGUIDialogKeyboardGeneric```, as it does with the result, and this dialog will launch the Toast

Reply:
Okay, but design the message format to work with enum  values, not directly with localized strings, please. Localized strings are also gui stuff that does not belong in the speech resolver listener class. I know this is a bit more work, but then we will end with a clean solution.

Reply:
I would define a new message

`#define TMSG_SPEECH_RECOGNIZER_ERROR  TMSG_MASK_APPLICATION + 34` in xbmc/messaging/ApplicationMesenger.h

and an 

`enum class SpeechRecognizerError` for the possible error values, in a new separate header file in xbmc/speech/SpeechRecognizerErrors.h (speech would be a new subdir)

`CJNISpeechRecognizer::_onError` uses AppMessenger to post the new message, which dispatches to `CApplication::OnApplicationMessage` which calls new method `CGUIDialogKeyboardGeneric::OnSpeechRecognitionError` which maps the enum value to localisable strings and shows the toast.